README.md
# PomodoroPlusPlus

## Description and Features

**PomodoroPlusPlus** is a standalone app with the goal of helping someone focus with the pomodoro method.

Even though it has a simple UI, the usage of the app is 100% free and intuitive for everyone to use.

## Content

- [PomodoroPlusPlus](#pomodoroplusplus)
  - [Description and Features](#description-and-features)
  - [Content](#content)
  - [Installation](#installation)
  - [Options](#options)
  - [Changelog](#changelog)

## Installation

1. Go to the [releases](https://github.com/thespbgamer/PomodoroPlusPlus/releases)
2. Download the `.exe` file
3. Run the `.exe` file

## Options

- Can change the following things
  - Time for work and rest
  - Number of sessions
  - Audio levels
  - Audio for certain actions (after rest, after work, finish and countdown)
  - Theme (dark and light)

## Changelog

- [Full Changelog](https://github.com/thespbgamer/PomodoroPlusPlus/blob/main/CHANGELOG.MD#full-changelog)


CHANGELOG.MD
# Full Changelog

## [1.2.0] - 2023-11-12

### Updated

- Dependencies
- Major logic optimizations
- More accurate timer

### Fixed

- Missing Assets (App & System Tray icons)

## [1.1.0] - 2023-09-23

### Added

- Close and minimize to the system tray.

### Improvements

- More accurate timers
- Layout and appearance overhaul
  - More responsive
  - Better appearance
  - More intuitive design
- Code optimization
- Packages optimizations

## [1.0.0] - 2023-08-03

### Added

- Initial release
  - Basic pomodoro and settings
    - It's possible to change the time durations, number os sessions, audio levels, audio files and program theme.

[Unreleased]: https://github.com/thespbgamer/PomodoroPlusPlus/compare/v1.2.0...HEAD
[1.2.0]: https://github.com/thespbgamer/PomodoroPlusPlus/releases/tag/v1.2.0
[1.1.0]: https://github.com/thespbgamer/PomodoroPlusPlus/releases/tag/v1.1.0
[1.0.0]: https://github.com/thespbgamer/PomodoroPlusPlus/releases/tag/v1.0.0


main.js
//npx tailwindcss -i ./src/assets/css/styles.css -o ./src/assets/css/tailwind.css --watch
require("dotenv").config();
const { app, BrowserWindow, ipcMain, nativeTheme, Tray, Menu } = require("electron");
const path = require("path");
let tray = null;

const isDev = process.env.NODE_ENV == "development";

const filesToCall = {
	"mainPreload.js": "./src/assets/js/globalPreload.js",
	"mainIndex.js": "./src/index.html"
};

function createWindow() {
	const win = new BrowserWindow({
		width: !isDev ? 1280 : 1500,
		height: 720,
		icon: "icon.ico",
		autoHideMenuBar: true,
		center: true,
		webPreferences: {
			preload: path.join(__dirname, filesToCall["mainPreload.js"]),
			nodeIntegration: true,
			backgroundThrottling: false,
			devTools: isDev
		}
	});

	win.on("minimize", function (event) {
		event.preventDefault();
		win.hide();
	});

	win.on("close", function (event) {
		if (!app.isQuiting) {
			event.preventDefault();
			win.hide();
		}
	});

	if (isDev) {
		win.webContents.openDevTools();
	}

	win.loadFile(filesToCall["mainIndex.js"]);

	tray = new Tray("icon.ico");
	const contextMenu = Menu.buildFromTemplate([
		{
			label: "Show",
			click: () => {
				win.show();
			}
		},
		{
			label: "Quit",
			click: () => {
				app.isQuiting = true;
				app.quit();
				functionQuit();
			}
		}
	]);
	tray.setToolTip("Pomodoro++");
	tray.setContextMenu(contextMenu);
	tray.addListener("double-click", function () {
		if (!win.isVisible()) {
			win.show();
		}
	});
}

ipcMain.handle("dark-mode:toggle", () => {
	if (nativeTheme.shouldUseDarkColors) {
		nativeTheme.themeSource = "light";
	} else {
		nativeTheme.themeSource = "dark";
	}
	return nativeTheme.shouldUseDarkColors;
});

ipcMain.handle("dark-mode:get", () => {
	return nativeTheme.shouldUseDarkColors ? "dark" : nativeTheme.shouldUseSystemColors;
});

app.whenReady().then(() => {
	createWindow();

	app.on("activate", () => {
		if (BrowserWindow.getAllWindows().length === 0) {
			createWindow();
		}
	});
});

app.on("window-all-closed", () => {});

function functionQuit() {
	if (process.platform !== "darwin") {
		app.quit();
	}
}


src/assets/js/globalPreload.js
const { contextBridge, ipcRenderer } = require("electron");
const fs = require("fs");
const path = require("path");
const sqlite3 = require("sqlite3");
const { open } = require("sqlite");

let database;

contextBridge.exposeInMainWorld("databaseOperations", {
	openDatabase: () => {
		if (!database) {
			database = open({
				filename: process.env.APPDATA + "\\PomodoroPlusPlus\\Data\\database.db",
				driver: sqlite3.cached.Database
			});
		}

		return database;
	}
});

//file management
contextBridge.exposeInMainWorld("files", {
	copy: (source, fileOutputName, fileToDelete) => {
		if (fs.existsSync(fileToDelete)) {
			fs.unlinkSync(fileToDelete);
		}

		if (!fs.existsSync(source)) {
			return;
		}

		if (fileOutputName == null || fileOutputName == undefined) {
			fileOutputName = Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
		}

		let sourceExtension = path.extname(source);
		//appdata path
		// console.log(process.env.APPDATA);
		let fileFinalLocation = process.env.APPDATA + "\\PomodoroPlusPlus\\Data\\" + fileOutputName + sourceExtension;

		//create folder if not exists
		if (!fs.existsSync(process.env.APPDATA + "\\PomodoroPlusPlus")) {
			fs.mkdirSync(process.env.APPDATA + "\\PomodoroPlusPlus");
		}
		if (!fs.existsSync(process.env.APPDATA + "\\PomodoroPlusPlus\\Data")) {
			fs.mkdirSync(process.env.APPDATA + "\\PomodoroPlusPlus\\Data");
		}

		fs.copyFileSync(source, fileFinalLocation);

		return fileFinalLocation;
	},
	doesExist: (filename) => {
		return fs.existsSync(process.env.APPDATA + "\\PomodoroPlusPlus\\Data\\" + filename);
	}
});

contextBridge.exposeInMainWorld("darkMode", {
	toggle: () => ipcRenderer.invoke("dark-mode:toggle"),
	get: () => ipcRenderer.invoke("dark-mode:get")
	// system: () => ipcRenderer.invoke("dark-mode:system"),
});


src/assets/js/indexRenderer.js
//Default invalid values for certain global variables
let glob_timeLeft = -1;
let glob_currentActivity = "N/A";
let glob_numberOfSessions = -1;
let glob_pomodoroTimerPartsRemaining = -1;
let glob_currentAudio;

//On page load
window.addEventListener("load", () => {
	//const db = window.databaseOperations.openDatabase();

	//If it's the first time opening the app, redirect to the options
	if (localStorage.getItem("numberOfSessionsValue") == null || localStorage.getItem("numberOfSessionsValue") < 1) {
		window.location.replace("options.html");
	}
	//Get the number of sessions and save it
	glob_numberOfSessions = localStorage.getItem("numberOfSessionsValue");
});

//When you click the start button
document.getElementById("startButton").addEventListener("click", () => {
	//Calculate the number of pomodoro timers remaining
	glob_pomodoroTimerPartsRemaining = glob_numberOfSessions * 2;

	//Force the current activity to "Work"
	toggleCurrentActivityAndTimeLeft("Work");

	//Start the countdown
	startCountdown();
});

function startCountdown() {
	//Has the countdown started ?
	let hasCountdownStarted = false;

	// Timers left -1
	glob_pomodoroTimerPartsRemaining -= 1;

	//If no more timers, define the pomodoro with "Finished", and re-enable the start button
	if (glob_pomodoroTimerPartsRemaining < 0) {
		updateCurrentSessionMessage(0, "Finish");
		toggleCurrentActivityAndTimeLeft("Rest");
		return;
	}

	//Disable the start button and update colors of text and the circle timer
	document.getElementById("startButton").setAttribute("disabled", "disabled");
	updateCurrentSessionMessage();

	const currentPartTimer = setInterval(function () {
		//format mm:ss with allwas 2 digits
		document.getElementById("Countdown").innerHTML =
			("0" + Math.floor(glob_timeLeft / 60)).slice(-2) + ":" + ("0" + (glob_timeLeft % 60)).slice(-2);
		glob_timeLeft -= 1;
		if (glob_timeLeft < 0) {
			if (glob_pomodoroTimerPartsRemaining == 0) {
				document.getElementById("Countdown").innerHTML = "Congratulations";
				playAudio("Finish");
			} else {
				document.getElementById("Countdown").innerHTML = "Next part is Starting";
				playAudio("Done");
			}
			clearInterval(currentPartTimer);

			document.getElementById("startButton").removeAttribute("disabled");
			toggleCurrentActivityAndTimeLeft();
			updateCurrentSessionMessage(0, "resetStyles");
			startCountdown();
		}
		if (glob_timeLeft < 3 && !hasCountdownStarted) {
			hasCountdownStarted = true;
			//play wav file
			playAudio("Countdown");
		}
	}, 1000);
}

function toggleCurrentActivityAndTimeLeft(forceValue = null) {
	if (glob_currentActivity == "Work" || forceValue == "Rest") {
		glob_currentActivity = "Rest";
		glob_timeLeft = localStorage.getItem("timeForPomodoroRestingSession");
	} else if (glob_currentActivity == "Rest" || forceValue == "Work") {
		glob_currentActivity = "Work";
		glob_timeLeft = localStorage.getItem("timeForPomodoroWorkingSession");
	}
}

function updateCurrentSessionMessage(delayInSeconds = 1, status = glob_currentActivity) {
	setTimeout(function () {
		resetColorClassesCurrentSession();

		if (status == "Rest") {
			updateCurrentTimerTextAndColor("Resting", "text-green-500");
			updateProgressBarColor("green", "drop-shadow(0px 0px 3px rgba(0, 255, 0, 0.7))");
		} else if (status == "Work") {
			updateCurrentTimerTextAndColor("Working", "text-red-600");
			updateProgressBarColor("red", "drop-shadow(0px 0px 3px rgba(255, 0, 0, 0.7))");
		} else if (status == "Finish") {
			updateCurrentTimerTextAndColor("Finished", "text-amber-400");
			updateProgressBarColor("gold", "drop-shadow(0px 0px 3px rgba(255,215,0, 0.7))");
		} else {
			//remove style color
			document.getElementById("currentSession").innerHTML = "Done";
			document.querySelector(".progressBarValue").style.setProperty("--animationTime", "0ms");
		}
	}, delayInSeconds * 1000);
}

function updateCurrentTimerTextAndColor(text, colorClass) {
	document.getElementById("currentSession").innerHTML = text;
	document.getElementById("currentSession").classList.add(colorClass);
}

function resetColorClassesCurrentSession() {
	//remove text danger class
	document.getElementById("currentSession").classList.remove("text-red-600");
	//remove text success class
	document.getElementById("currentSession").classList.remove("text-green-500");
	//remove golden text
	document.getElementById("currentSession").classList.remove("text-amber-400");
}

function updateProgressBarColor(barColor, barDropShadow, timeValue = glob_timeLeft) {
	document.querySelector(".progressBarValue").style.setProperty("--animationTime", "");

	document.querySelector(".progressBarValue").style.setProperty("stroke", barColor);
	document.querySelector(".progressBarValue").style.setProperty("filter", barDropShadow);

	document.querySelector(".progressBarValue").style.setProperty("--animationTime", timeValue + "s");
	document.querySelector(".progressBarValue").style.setProperty("animation", "none");
	setTimeout(function () {
		document.querySelector(".progressBarValue").style.setProperty("animation", "");
	}, 5);
}

function playAudio(currentAudioToPlay) {
	if (glob_currentAudio != null && glob_currentAudio != undefined) {
		glob_currentAudio.pause();
	}
	const audioName = currentAudioToPlay != "Done" ? currentAudioToPlay : glob_currentActivity;

	const audioPath = "audio" + audioName + "Path";

	if (localStorage.getItem(audioPath) != null) {
		glob_currentAudio = new Audio(localStorage.getItem(audioPath));
	} else if (currentAudioToPlay == "Done") {
		glob_currentAudio = new Audio("assets/audio/" + glob_currentActivity + ".wav");
	} else {
		glob_currentAudio = new Audio("assets/audio/" + currentAudioToPlay + ".wav");
	}

	//audio level set to 100%
	glob_currentAudio.volume = localStorage.getItem("audioLevelValue") || 1;
	glob_currentAudio.play();
}


src/assets/js/optionsRenderer.js
let audioAfterWorkFile = null;
let audioAfterRestFile = null;
let audioCountDownFile = null;
let audioFinishFile = null;
let notificationCount;

document.addEventListener("keypress", function (e) {
	if (e.key === "Enter") {
		e.preventDefault();
		return false;
	}
});

document.getElementById("toggle-dark-mode").addEventListener("click", () => {
	const isDarkMode = window.darkMode.toggle();
	document.getElementById("bodyColor").setAttribute("data-bs-theme", isDarkMode ? "dark" : "light");
});

document.getElementById("audioAfterWorkFile").addEventListener("change", () => {
	audioAfterWorkFile = document.getElementById("audioAfterWorkFile").files[0];
});

document.getElementById("audioAfterRestFile").addEventListener("change", () => {
	audioAfterRestFile = document.getElementById("audioAfterRestFile").files[0];
});

document.getElementById("audioCountDownFile").addEventListener("change", () => {
	audioCountDownFile = document.getElementById("audioCountDownFile").files[0];
});

document.getElementById("audioFinishFile").addEventListener("change", () => {
	audioFinishFile = document.getElementById("audioFinishFile").files[0];
});

document.getElementById("reset-audio-after-work").addEventListener("click", () => {
	window.files.copy(null, null, localStorage.getItem("audioWorkPath"));
	localStorage.removeItem("audioWorkPath");
	audioAfterWorkFile = null;
	document.getElementById("audioAfterWorkFile").value = null;

	showAlertAndDismiss("alert-blue-div", "alert-blue-text", "Audio after work reset.", 3000, "blue");
});

document.getElementById("reset-audio-after-rest").addEventListener("click", () => {
	window.files.copy(null, null, localStorage.getItem("audioRestPath"));
	localStorage.removeItem("audioRestPath");
	audioAfterRestFile = null;
	document.getElementById("audioAfterRestFile").value = null;
	showAlertAndDismiss("alert-blue-div", "alert-blue-text", "Audio after rest reset.", 3000, "blue");
});

document.getElementById("reset-audio-countdown").addEventListener("click", () => {
	window.files.copy(null, null, localStorage.getItem("audioCountdownPath"));
	localStorage.removeItem("audioCountdownPath");
	audioCountDownFile = null;
	document.getElementById("audioCountDownFile").value = null;
	showAlertAndDismiss("alert-blue-div", "alert-blue-text", "Audio countdown reset.", 3000, "blue");
});

document.getElementById("reset-audio-finish").addEventListener("click", () => {
	window.files.copy(null, null, localStorage.getItem("audioFinishPath"));
	localStorage.removeItem("audioFinishPath");
	audioFinishFile = null;
	document.getElementById("audioFinishFile").value = null;
	showAlertAndDismiss("alert-blue-div", "alert-blue-text", "Audio finish reset.", 3000, "blue");
});

document.getElementById("save-options").addEventListener("click", () => {
	try {
		let timeForPomodoroWorkingSession = document.getElementById("timeForPomodoroWorkingSession").value;
		let timeForPomodoroRestingSession = document.getElementById("timeForPomodoroRestingSession").value;
		let numberOfSessionsValue = document.getElementById("numberOfSessionsValue").value;
		let audioLevelValue = document.getElementById("audioLevelValue").value;

		if (
			timeForPomodoroWorkingSession == "" ||
			timeForPomodoroWorkingSession <= 0 ||
			timeForPomodoroWorkingSession > 86400 ||
			timeForPomodoroRestingSession == "" ||
			timeForPomodoroRestingSession <= 0 ||
			timeForPomodoroRestingSession > 86400 ||
			numberOfSessionsValue == "" ||
			numberOfSessionsValue <= 0 ||
			numberOfSessionsValue > 100 ||
			audioLevelValue == "" ||
			audioLevelValue < 0 ||
			audioLevelValue > 100
		) {
			return;
		}

		localStorage.setItem("timeForPomodoroWorkingSession", timeForPomodoroWorkingSession);
		localStorage.setItem("timeForPomodoroRestingSession", timeForPomodoroRestingSession);
		localStorage.setItem("numberOfSessionsValue", numberOfSessionsValue);
		localStorage.setItem("audioLevelValue", audioLevelValue / 100);

		if (audioAfterWorkFile != null && audioAfterWorkFile != undefined) {
			let finalDestination = window.files.copy(audioAfterWorkFile.path, null, localStorage.getItem("audioWorkPath"));
			localStorage.setItem("audioWorkPath", finalDestination);
		}

		if (audioAfterRestFile != null && audioAfterRestFile != undefined) {
			let finalDestination = window.files.copy(audioAfterRestFile.path, null, localStorage.getItem("audioRestPath"));
			localStorage.setItem("audioRestPath", finalDestination);
		}

		if (audioCountDownFile != null && audioCountDownFile != undefined) {
			let finalDestination = window.files.copy(
				audioCountDownFile.path,
				null,
				localStorage.getItem("audioCountdownPath")
			);
			localStorage.setItem("audioCountdownPath", finalDestination);
		}

		if (audioFinishFile != null && audioFinishFile != undefined) {
			let finalDestination = window.files.copy(audioFinishFile.path, null, localStorage.getItem("audioFinishPath"));
			localStorage.setItem("audioFinishPath", finalDestination);
		}

		showAlertAndDismiss("alert-green-div", "alert-green-text", "Options Saved!", 3000, "green");
	} catch (err) {
		showAlertAndDismiss("alert-red-div", "alert-red-text", err, 3000, "red");
	}
});

window.addEventListener("load", () => {
	notificationCount = 0;
	let timeForPomodoroWorkingSession = localStorage.getItem("timeForPomodoroWorkingSession");
	let timeForPomodoroRestingSession = localStorage.getItem("timeForPomodoroRestingSession");
	let numberOfSessionsValue = localStorage.getItem("numberOfSessionsValue");
	let audioLevelValue = localStorage.getItem("audioLevelValue");

	if (numberOfSessionsValue == null || parseInt(numberOfSessionsValue) < 1) {
		numberOfSessionsValue = 3;
		showAlertAndDismiss("alert-blue-div", "alert-blue-text", "Confirm the options and click save.", 3000, "purple");
	}

	if (audioLevelValue == null || parseInt(audioLevelValue) < 0 || parseInt(audioLevelValue) > 1) {
		audioLevelValue = 1;
	}

	if (timeForPomodoroWorkingSession == null) {
		timeForPomodoroWorkingSession = 1500;
	}

	if (timeForPomodoroRestingSession == null) {
		timeForPomodoroRestingSession = 900;
	}

	document.getElementById("timeForPomodoroWorkingSession").value = timeForPomodoroWorkingSession;
	document.getElementById("timeForPomodoroRestingSession").value = timeForPomodoroRestingSession;
	document.getElementById("numberOfSessionsValue").value = numberOfSessionsValue;
	document.getElementById("audioLevelValue").value = audioLevelValue * 100;

	convertTimerValues("timeForPomodoroWorkingSession", "timeToWorkConverted");
	convertTimerValues("timeForPomodoroRestingSession", "timeToRestConverted");
	convertTimerValues("numberOfSessionsValue", "finalTimeConverted");
});

document.getElementById("timeForPomodoroWorkingSession").addEventListener("input", () => {
	convertTimerValues("timeForPomodoroWorkingSession", "timeToWorkConverted");
	convertTimerValues("numberOfSessionsValue", "finalTimeConverted");
});

document.getElementById("timeForPomodoroRestingSession").addEventListener("input", () => {
	convertTimerValues("timeForPomodoroRestingSession", "timeToRestConverted");
	convertTimerValues("numberOfSessionsValue", "finalTimeConverted");
});

document.getElementById("numberOfSessionsValue").addEventListener("input", () => {
	convertTimerValues("numberOfSessionsValue", "finalTimeConverted");
});

document.getElementById("play-audio-after-work").addEventListener("click", () => {
	playAudio("Done", "Work", audioAfterWorkFile?.path);
});

document.getElementById("play-audio-after-rest").addEventListener("click", () => {
	playAudio("Done", "Rest", audioAfterRestFile?.path);
});
document.getElementById("play-audio-countdown").addEventListener("click", () => {
	playAudio("Countdown", null, audioCountDownFile?.path);
});

document.getElementById("play-audio-finish").addEventListener("click", () => {
	playAudio("Finish", null, audioFinishFile?.path);
});

function convertTimerValues(originalID, convertedID) {
	let timeForPomodoroWorkingSession = document.getElementById(originalID).value;

	if (timeForPomodoroWorkingSession == "" || timeForPomodoroWorkingSession.length == 0) {
		document.getElementById(convertedID).innerHTML = "N/A";
		return;
	} else if (convertedID == "finalTimeConverted") {
		timeForPomodoroWorkingSession =
			(parseInt(document.getElementById("timeForPomodoroWorkingSession").value) +
				parseInt(document.getElementById("timeForPomodoroRestingSession").value)) *
			parseInt(document.getElementById("numberOfSessionsValue").value);
	}

	//get hours
	let hoursLeft = Math.floor(timeForPomodoroWorkingSession / 3600);

	//get minutes
	let minutesLeft = Math.floor(timeForPomodoroWorkingSession / 60) % 60;

	//get remaining seconds
	let secondsLeft = timeForPomodoroWorkingSession % 60;

	// Padding the values to ensure they are two digits
	if (hoursLeft < 10) {
		hoursLeft = "0" + hoursLeft;
	}
	if (minutesLeft < 10) {
		minutesLeft = "0" + minutesLeft;
	}
	if (secondsLeft < 10) {
		secondsLeft = "0" + secondsLeft;
	}

	document.getElementById(convertedID).innerHTML =
		secondsLeft >= 0 ? hoursLeft + ":" + minutesLeft + ":" + secondsLeft : "N/A";
}

let currentAudio;

function playAudio(currentAudioToPlay, currentActivity = "", audioOverwrite = null) {
	if (currentAudio != null && currentAudio != undefined) {
		currentAudio.pause();
	}

	const audioName = !currentActivity ? currentAudioToPlay : currentActivity;

	const audioPath = "audio" + audioName + "Path";

	if (audioOverwrite) {
		currentAudio = new Audio(audioOverwrite);
	} else if (localStorage.getItem(audioPath) != null) {
		currentAudio = new Audio(localStorage.getItem(audioPath));
	} else if (currentAudioToPlay == "Done") {
		currentAudio = new Audio("assets/audio/" + currentActivity + ".wav");
	} else {
		currentAudio = new Audio("assets/audio/" + currentAudioToPlay + ".wav");
	}

	//audio level set to 100%
	currentAudio.volume = localStorage.getItem("audioLevelValue") || 1;
	currentAudio.play();
}

function showAlertAndDismiss(alertDivID, alertTextID, alertText, alertDismissTimeInMS = 3000, color = "blue") {
	let notificationArea = document.getElementById("notificationArea");

	if (notificationArea.innerHTML.length <= 0) {
		notificationCount = 0;
	}
	notificationCount += 1;

	let currentDivID = alertDivID + notificationCount;
	let currentTextID = alertTextID + notificationCount;

	let currentDismiss = alertDismissTimeInMS * /* notificationCount */ 1;

	let innerHTMLForDiv =
		`<div
			id="` +
		currentDivID +
		`"
		class="unselectable opacity-0 duration-200 ease-out transition transform fixed bottom-0 right-2 flex items-center p-4 z-100 mb-24 text-` +
		color +
		`-800 rounded-lg bg-` +
		color +
		`-50 dark:bg-` +
		color +
		`-950 dark:text-` +
		color +
		`-400"
		role="alert">
		<i class="bi bi-info-circle-fill"></i>
		<span class="sr-only">Info</span>
		<div id="` +
		currentTextID +
		`" class="ml-1 text-sm font-medium">
			</div>
		</div>`;

	notificationArea.innerHTML = innerHTMLForDiv;

	setTimeout(() => {
		if (document.getElementById(currentDivID) != null) {
			document.getElementById(currentDivID).classList.remove("opacity-0");
			document.getElementById(currentTextID).innerHTML = alertText;

			setTimeout(() => {
				if (document.getElementById(currentDivID) != null) {
					document.getElementById(currentDivID).classList.add("opacity-0");
					setTimeout(() => {
						if (document.getElementById(currentDivID) != null) {
							document.getElementById(currentDivID).remove();
						}
					}, 250);
				}
			}, currentDismiss);
		}
	}, 50);
}
